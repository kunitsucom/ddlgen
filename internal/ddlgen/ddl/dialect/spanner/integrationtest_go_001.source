package main

type (
	// User is a user.
	//
	// spanddl:      table: `Users`
	// spanddl: constraint: CONSTRAINT AgeGTEZero CHECK(Age >= 0)
	// spanddl:      index: CREATE INDEX `IndexUsersByName` ON `Users` (`Name`)
	User struct {
		// UserID is a user ID.
		UserID string `dbtest:"UserId" spanddl:"STRING(36)  NOT NULL" pkey:"true"`
		// Name is a user name.
		Name string   `dbtest:"Name"   spanddl:"STRING(255) NOT NULL"`
		// Email is a user email.
		Email string  `dbtest:"Email"  spanddl:"STRING(255) NOT NULL"`
		// Age is a user age.
		Age int       `dbtest:"Age"    spanddl:"INT64       NOT NULL"`
		// Ignore is a ignore field.
		Ignore string `dbtest:"-"      spanddl:"-"`
	}

	// Users is a user array.
	// This type is expected not to be detected.
	//
	// spanddl: table: `UserArrays`
	Users []*User

	// UserGroup is a user group.
	//
	// spanddl:table:
	UserGroup struct {
		// ID is a group ID.
		ID string   `dbtest:"Id"   spanddl:"STRING(36)  NOT NULL"`
		// Name is a group name.
		Name string `dbtest:"Name" spanddl:"STRING(255) NOT NULL"`
	}

	// Author is a author.
	// spanddl:
	Author struct {
		// ID is a author ID.
		AuthorID string `dbtest:"Id"     spanddl:"STRING(36)  NOT NULL"`
		// Name is a author name.
		Name string     `dbtest:"Name"   spanddl:"STRING(255) NOT NULL"`
	}

	// Book is a book.
	//
	//spanddl:table `Books`
	//spanddl:option INTERLEAVE IN PARENT `Author` ON DELETE CASCADE
	Book struct {
		// AuthorID is a book author.
		AuthorID string `spanddl:"STRING(255) NOT NULL" pkey:"true"`
		// ID is a book ID.
		ID string       `spanddl:"STRING(36)  NOT NULL" pkey:"true"`
		// Title is a book title.
		Title string    `spanddl:"STRING(255) NOT NULL"`
	}

	// Store
	// spanddl: table: CREATE TABLE `Stores`
	Store struct {
		// ID is a store ID.
		ID string   `dbtest:"Id" pkey:"false"`
		// Name is a store name.
		Name string `dbtest:"Name"`
	}
)

// spanddl: index: `IndexBooksByTitle` ON `Books` (`Title`)
