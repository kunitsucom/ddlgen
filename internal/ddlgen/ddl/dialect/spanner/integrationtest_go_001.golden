-- Code generated by ddlgen. DO NOT EDIT.
--
-- Date: 2021-01-01T09:00:00+09:00
--

-- source: spanner/integrationtest_go_001.source:4
-- User is a user.
--
-- spanddl:      table: `Users`
-- spanddl:     option: PRIMARY KEY (`Id`)
-- spanddl: constraint: CONSTRAINT AgeGTEZero CHECK(Age >= 0)
-- NOTE: the "User" struct's field "Ignore" has a tag for column name (`dbtest:"-"`), so the field is ignored.
CREATE TABLE `Users` (
    -- ID is a user ID.
    `Id`    STRING(36)  NOT NULL,
    -- Name is a user name.
    `Name`  STRING(255) NOT NULL,
    -- Email is a user email.
    `Email` STRING(255) NOT NULL,
    -- Age is a user age.
    `Age`   INT64       NOT NULL,
    CONSTRAINT AgeGTEZero CHECK(Age >= 0)
)
PRIMARY KEY (`Id`);

-- source: spanner/integrationtest_go_001.source:9
-- spanddl:      index: CREATE INDEX `IndexUsersByName` ON Users(`Name`)
CREATE INDEX `IndexUsersByName` ON Users(`Name`);

-- source: spanner/integrationtest_go_001.source:23
-- Users is a user array.
-- This type is expected not to be detected.
--
-- spanddl: table: `UserArrays`
-- ERROR: the comment (spanner/integrationtest_go_001.source:23) does not have struct fields for column type and constraint (`spanddl:"<TYPE> [CONSTRAINT]"`), or the comment is not associated with struct.
CREATE TABLE `UserArrays` (
);

-- source: spanner/integrationtest_go_001.source:29
-- Group is a group.
--
-- spanddl:table:
-- WARN: the comment (spanner/integrationtest_go_001.source:29) does not have a key for table (spanddl: table: CREATE TABLE <table>), so the struct name "Group" is used as the table name.
CREATE TABLE Group (
    -- ID is a group ID.
    `Id`   STRING(36)  NOT NULL,
    -- Name is a group name.
    `Name` STRING(255) NOT NULL
);

-- source: spanner/integrationtest_go_001.source:39
-- Book is a book.
--
-- spanddl:table:`Books`
-- spanddl:options:PRIMARY KEY (`Id`)
CREATE TABLE `Books` (
    -- NOTE: the struct does not have a tag for column name (`dbtest:"<ColumnName>"`), so the field name "ID" is used as the column name.
    -- ID is a book ID.
    `ID`     STRING(36)  NOT NULL,
    -- NOTE: the struct does not have a tag for column name (`dbtest:"<ColumnName>"`), so the field name "Title" is used as the column name.
    -- Title is a book title.
    `Title`  STRING(255) NOT NULL,
    -- NOTE: the struct does not have a tag for column name (`dbtest:"<ColumnName>"`), so the field name "Author" is used as the column name.
    -- Author is a book author.
    `Author` STRING(255) NOT NULL
)
PRIMARY KEY (`Id`);

-- source: spanner/integrationtest_go_001.source:52
-- Author is a author.
-- spanddl:
-- WARN: the comment (spanner/integrationtest_go_001.source:52) does not have a key for table (spanddl: table: CREATE TABLE <table>), so the struct name "Author" is used as the table name.
CREATE TABLE Author (
    -- ID is a author ID.
    `Id`   STRING(36)  NOT NULL,
    -- Name is a author name.
    `Name` STRING(255) NOT NULL
);

-- source: spanner/integrationtest_go_001.source:61
-- Store
-- spanddl: table: CREATE TABLE `Stores`
CREATE TABLE `Stores` (
    -- ERROR: the struct does not have a tag for column type and constraint (`spanddl:"<TYPE> [CONSTRAINT]"`)
    -- ID is a store ID.
    `Id`   ERROR,
    -- ERROR: the struct does not have a tag for column type and constraint (`spanddl:"<TYPE> [CONSTRAINT]"`)
    -- Name is a store name.
    `Name` ERROR
);

-- source: spanner/integrationtest_go_001.source:71
-- spanddl: index: CREATE INDEX `IndexBooksByTitle` ON Books(`Title`)
CREATE INDEX `IndexBooksByTitle` ON Books(`Title`);

-- source: spanner/integrationtest_go_001.source:71
-- WARN: the comment (spanner/integrationtest_go_001.source:71) does not have a key for table (spanddl: table: CREATE TABLE <table>), or the comment is not associated with struct.
